httpbp/client_middlewares.go

```diff
@@ -0,0 +1,239 @@
+package httpbp
+
+import (
+	"net/http"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/opentracing/opentracing-go"
+	"github.com/reddit/baseplate.go/breakerbp"
+	"github.com/reddit/baseplate.go/tracing"
+	"github.com/sony/gobreaker"
+)
+
+// ClientMiddleware is used to build HTTP client middlewares by implementing
+// http.RoundTripper which http.Client accepts as Transport.
+type ClientMiddleware func(roundTripper http.RoundTripper) http.RoundTripper
+
+// NewClient returns a standard HTTP client wrapped with the default
+// middlewares plus any additional client middlewares passed into this
+// function.
+func NewClient(config Config, middlewares ...ClientMiddleware) *http.Client {
+	transport := &http.Transport{
+		MaxConnsPerHost: config.MaxConnections,
+	}
+	defaults := []ClientMiddleware{
+		MonitorClient(config),
+		CircuitBreaker(config.CircuitBreaker),
+		Retries(config),
+	}
+	middlewares = append(middlewares, defaults...)
+
+	return &http.Client{
+		Transport: applyMiddlewares(transport, middlewares...),
+	}
+}
+
+func applyMiddlewares(transport http.RoundTripper, middlewares ...ClientMiddleware) http.RoundTripper {
+	// add middlewares in reverse so the first in the list is the outermost
+	for i := len(middlewares) - 1; i >= 0; i-- {
+		transport = middlewares[i](transport)
+	}
+	return transport
+}
+
+type circuitBreaker struct {
+	next     http.RoundTripper
+	settings gobreaker.Settings
+
+	mu       sync.RWMutex
+	breakers map[string]breakerbp.CircuitBreaker
+}
+
+// CircuitBreaker is a middleware that prevents sending requests that are likely
+// to fail based on a configurable failure ratio based on total failures and
+// requests.
+func CircuitBreaker(config CircuitBreakerConfig) ClientMiddleware {
+	settings := gobreaker.Settings{
+		Interval: 60 * time.Second, // Reset the counts every 60 seconds
+		ReadyToTrip: func(counts gobreaker.Counts) bool {
+			failureRatio := float64(counts.TotalFailures) / float64(counts.Requests)
+			return counts.Requests >= config.MinRequests && failureRatio >= config.MinFailureRatio
+		},
+	}
+	return func(next http.RoundTripper) http.RoundTripper {
+		return &circuitBreaker{
+			next:     next,
+			settings: settings,
+		}
+	}
+}
+
+func (c *circuitBreaker) RoundTrip(req *http.Request) (*http.Response, error) {
+	host := req.URL.Hostname()
+
+	c.mu.RLock()
+	breaker, ok := c.breakers[host]
+	if !ok {
+		c.setupBreaker(host)
+	}
+	c.mu.RUnlock()
+
+	resp, err := breaker.Execute(c.roundTrip(req))
+	if err != nil {
+		return nil, err
+	}
+	return resp.(*http.Response), nil
+}
+
+func (c *circuitBreaker) setupBreaker(host string) {
+	// this code path is executed rarely and usually at the start of a process
+	c.mu.RUnlock()
+	c.mu.Lock()
+	c.breakers[host] = gobreaker.NewCircuitBreaker(c.settings)
+	c.mu.Unlock()
+	c.mu.RLock()
+}
```

Comment:
Can someone think of a better way than this locking Shashlik? The circuit breaker should be able to support a circuit breaker per host being queried and I can see how it makes sense to have it initialize lazy. In most cases we just want a read-lock and this is the path for initializing it requiring a write lock.

Reply:
I wrote https://pkg.go.dev/go.yhsif.com/defaultdict, which is basically pairing a `sync.Map` with `sync.Pool`.

Reply:
but then again, do we really need this breaker to be per-host?

we also have `slug` for the client, which kinds of imply that we don't really use a single client for everything (like the default `http.Client`), but we actually create a client for every upstream. In that case, a single breaker should be sufficient?

Reply:
Possibly, the one scenario I can think of here is the load balancer, i.e. we are talking to different Solr read replicas and we'd want the breaker to sort out a faulty endpoint.